src/picongpu/include/particles/ionization/byField/BSI/BSI.hpp

```diff
@@ -39,23 +50,157 @@ namespace ionization
+                 * \tparam T_BBox type of B-field box
+                 * \tparam T_EBox type of E-field box
+                 *
+                 * \param ionFrame (here address of: ) frame of the to-be-ionized particles
+                 * \param localIdx local (linear) index in super cell / frame
+                 * \param bBox B-field box instance
+                 * \param ebox E-field box instance
+                 * \param newMacroElectrons (here address of: ) variable for each thread that stores the number
+                 *        of macro electrons to be created during the current time step
+                 */
+                template<class FrameType, class T_BBox, class T_EBox >
+                DINLINE void operator()(FrameType& ionFrame, int localIdx, T_BBox& bBox, T_EBox& eBox, unsigned int& newMacroElectrons)
+                {
+
+                    /** type for field to particle interpolation */
+                    typedef T_Field2ParticleInterpolation Field2ParticleInterpolation;
+
+                    typedef typename T_BBox::ValueType BType;
+                    typedef typename T_EBox::ValueType EType;
+                    /* alias for the single macro-particle */
+                    PMACC_AUTO(particle,ionFrame[localIdx]);
+
+                    floatD_X pos = particle[position_];
+                    const int particleCellIdx = particle[localCellIdx_];
+
+                    DataSpace<TVec::dim> localCell(DataSpaceOperations<TVec::dim>::template map<TVec > (particleCellIdx));
+
+                    EType eField = Field2ParticleInterpolation()
+                        (eBox.shift(localCell).toCursor(), pos, NumericalCellType::getEFieldPosition());
+
+                    BType bField =Field2ParticleInterpolation()
+                        (bBox.shift(localCell).toCursor(), pos, NumericalCellType::getBFieldPosition());
+
+                    /* define number of bound macro electrons before ionization */
+                    float_X prevBoundElectrons = particle[boundElectrons_];
+
+                    /* this is the point where actual ionization takes place */
+                    IonizeAlgo ionizeAlgo;
+                    ionizeAlgo(
+                         bField, eField,
+                         particle
+                         );
+
+                    /* determine number of new macro electrons to be created */
+                    newMacroElectrons = prevBoundElectrons - particle[boundElectrons_];
+                    /*calculate one dimensional cell index*/
+
+                }
+            }; // struct IonizeParticlesPerFrame
+
+
+            /** Functor implementation
+             * \tparam T_SrcSpecies particle species to be ionized
+             * \tparam T_ParticleStorage \see picongpu/src/libPMacc/include/math/MapTuple.hpp
+             */
+            template<typename T_SrcSpecies,typename T_ParticleStorage>
+            HINLINE void operator()(T_SrcSpecies& src, T_ParticleStorage& pst)
```

Comment:
But since the kernel cannot know about a random generator, the implementation of `FrameIonizer` should. That again means, though, that the implementation is different for BSI and ADK.